src/history/hs_verify.c

```diff
@@ -0,0 +1,230 @@
+        if (btree_id != this_btree_id)
+            break;
+
+        /*
+         * If we have already checked against this key, keep going to the next key. We only need to
+         * check the key once.
+         */
+        WT_ERR(__wt_compare(session, NULL, &key, prev_key, &cmp));
+        if (cmp == 0)
+            continue;
+
+        /* Check the key can be found in the data store.*/
+        if (CUR2BT(ds_cbt)->type == BTREE_ROW) {
+            WT_WITH_PAGE_INDEX(
+              session, ret = __wt_row_search(ds_cbt, &key, false, NULL, false, NULL));
+        } else {
+            up = (const uint8_t *)key.data;
+            WT_ERR(__wt_vunpack_uint(&up, key.size, &recno));
+            WT_WITH_PAGE_INDEX(session, ret = __wt_col_search(ds_cbt, recno, NULL, false, NULL));
+        }
+        WT_ERR(ret);
+
+        if (ds_cbt->compare != 0) {
+            F_SET(S2C(session), WT_CONN_DATA_CORRUPTION);
+            WT_ERR_PANIC(session, WT_PANIC,
+              "the associated history store key %s was not found in the data store %s",
+              __wt_buf_set_printable(session, key.data, key.size, false, prev_key),
+              session->dhandle->name);
+        }
+
+        WT_ERR(__cursor_reset(ds_cbt));
+
+        /*
+         * Copy the key memory into our scratch buffer. The key will get invalidated on our next
+         * cursor iteration.
+         */
+        WT_ERR(__wt_buf_set(session, prev_key, key.data, key.size));
+    }
+    WT_ERR_NOTFOUND_OK(ret, true);
+err:
+    F_CLR(ds_cbt, WT_CURSTD_IGNORE_TOMBSTONE);
+    WT_ASSERT(session, key.mem == NULL && key.memsize == 0);
+    __wt_scr_free(session, &prev_key);
+    return (ret);
+}
+
+/*
+ * __wt_hs_verify_one --
+ *     Verify the history store for a given btree. This must be called when we are known to have
+ *     exclusive access to the btree.
+ */
+int
+__wt_hs_verify_one(WT_SESSION_IMPL *session, uint32_t this_btree_id)
+{
+    WT_CURSOR *hs_cursor;
+    WT_CURSOR_BTREE ds_cbt;
+    WT_DECL_RET;
+    WT_ITEM key;
+    wt_timestamp_t hs_start_ts;
+    uint64_t hs_counter;
+    uint32_t btree_id;
+
+    hs_cursor = NULL;
+    WT_CLEAR(key);
+    hs_start_ts = 0;
+    hs_counter = 0;
+    btree_id = WT_BTREE_ID_INVALID;
+
+    WT_ERR(__wt_curhs_open(session, NULL, &hs_cursor));
+    F_SET(hs_cursor, WT_CURSTD_HS_READ_COMMITTED);
+
+    /* Position the hs cursor on the requested btree id, there could be nothing in the HS yet. */
+    hs_cursor->set_key(hs_cursor, 1, this_btree_id);
+    WT_ERR_NOTFOUND_OK(__wt_curhs_search_near_after(session, hs_cursor), true);
+    if (ret == WT_NOTFOUND) {
+        ret = 0;
+        goto err;
+    }
+
+    /*
+     * Make sure the cursor is positioned on the given btree id, if not, there is nothing in the HS
+     * to be checked.
+     */
+    WT_ERR(hs_cursor->get_key(hs_cursor, &btree_id, &key, &hs_start_ts, &hs_counter));
+    if (this_btree_id != btree_id)
```

Comment:
From my understanding I think we might not need this check, I believe it would directly return back WT_NOTFOUND. Be good to double check this.

Reply:
When checking the code of `__wt_curhs_search_near_after`, it positions the cursor on the next greater key if the exact key is not found.
If I ask for the `btree_id` 1 and only records with `btree_id` 2 are present, the cursor will be positioned on `btree_id` 2. Since this is not the requested `btree_id`, we should exit here.

Reply:
After some testing, it seems that you are right but I don't fully understand why. It seems that `__wt_curhs_search_near_after` can only return a greater record that contains `this_btree_id`. If no records contain `this_btree_id`, then the function returns `WT_NOTFOUND`.

Reply:
Removed the unnecessary check in aa4cc8be94235853044e06af7531814f094429ca, thanks!